beacon/validator/src/test/java/tech/pegasys/teku/validator/coordinator/GraffitiBuilderTest.java

```diff
@@ -0,0 +1,424 @@
+            graffitiBuilder.extractGraffiti(
+                Optional.of(Bytes32Parser.toBytes32(asciiGraffiti0)), 0))
+        .isEqualTo("");
+    assertThat(
+            graffitiBuilder.extractGraffiti(
+                Optional.of(Bytes32Parser.toBytes32(ASCII_GRAFFITI_20)), 20))
+        .isEqualTo(ASCII_GRAFFITI_20);
+    assertThat(
+            graffitiBuilder.extractGraffiti(
+                Optional.of(Bytes32Parser.toBytes32(asciiGraffiti32)), 32))
+        .isEqualTo(asciiGraffiti32);
+    assertThat(
+            graffitiBuilder.extractGraffiti(
+                Optional.of(Bytes32Parser.toBytes32(asciiGraffiti32)), 16))
+        .isEqualTo(asciiGraffiti32.substring(0, 16));
+    assertThat(
+            graffitiBuilder.extractGraffiti(
+                Optional.of(Bytes32Parser.toBytes32(UTF_8_GRAFFITI_4)), 4))
+        .isEqualTo(UTF_8_GRAFFITI_4);
+    assertThat(
+            graffitiBuilder.extractGraffiti(
+                Optional.of(Bytes32Parser.toBytes32(utf8Graffiti20)), 20))
+        .isEqualTo(utf8Graffiti20);
+    assertThat(
+            graffitiBuilder.extractGraffiti(
+                Optional.of(Bytes32Parser.toBytes32(utf8Graffiti20)), 24))
+        .isEqualTo(utf8Graffiti20 + "\u0000\u0000\u0000\u0000");
+  }
+
+  @Test
+  public void calculateGraffitiLength_shouldReturnCorrectLength() {
+    assertThat(graffitiBuilder.calculateGraffitiLength(Optional.empty())).isEqualTo(0);
+    assertThat(
+            graffitiBuilder.calculateGraffitiLength(
+                Optional.of(Bytes32Parser.toBytes32(asciiGraffiti0))))
+        .isEqualTo(0);
+    assertThat(
+            graffitiBuilder.calculateGraffitiLength(
+                Optional.of(Bytes32Parser.toBytes32(ASCII_GRAFFITI_20))))
+        .isEqualTo(20);
+    assertThat(
+            graffitiBuilder.calculateGraffitiLength(
+                Optional.of(Bytes32Parser.toBytes32(asciiGraffiti32))))
+        .isEqualTo(32);
+    assertThat(
+            graffitiBuilder.calculateGraffitiLength(
+                Optional.of(Bytes32Parser.toBytes32(UTF_8_GRAFFITI_4))))
+        .isEqualTo(4);
+    assertThat(
+            graffitiBuilder.calculateGraffitiLength(
+                Optional.of(Bytes32Parser.toBytes32(utf8Graffiti20))))
+        .isEqualTo(20);
+  }
+
+  @Test
+  public void join_shouldJoinStringsSkippingEmpty() {
+    assertThat(graffitiBuilder.joinNonEmpty("", "aa", "bb", "cc"))
+        .isEqualTo(Bytes32Parser.toBytes32("aabbcc"));
+    assertThat(graffitiBuilder.joinNonEmpty(" ", "aa", "", "cc"))
+        .isEqualTo(Bytes32Parser.toBytes32("aa cc"));
+    assertThat(graffitiBuilder.joinNonEmpty(" ", "", "bb", "cc"))
+        .isEqualTo(Bytes32Parser.toBytes32("bb cc"));
+    assertThat(graffitiBuilder.joinNonEmpty(" ", "", "", "")).isEqualTo(Bytes32.ZERO);
+  }
+
+  @Test
+  public void formatClientInfo_shouldFitInDesiredLength() {
+    graffitiBuilder.onExecutionClientVersion(BESU_CLIENT_VERSION);
+
+    // 20: LH1be52536BU0f91a674
```

Comment:
we're really bundling multiple tests here, so if it ever breaks we'll only see the first failure

Reply:
ok, will split it